core/src/main/java/com/linecorp/armeria/common/stream/EventLoopStreamMessage.java

```diff
@@ -0,0 +1,417 @@
+        assert eventLoop.inEventLoop();
+        final AwaitDemandFuture future = new AwaitDemandFuture();
+        queue.add(future);
+        tryNotifyEvents();
+        return future;
+    }
+
+    @Override
+    public void close() {
+        if (eventLoop.inEventLoop()) {
+            close0(SUCCESSFUL_CLOSE);
+        } else {
+            eventLoop.execute(() -> close0(SUCCESSFUL_CLOSE));
+        }
+    }
+
+    @Override
+    public void close(Throwable cause) {
+        requireNonNull(cause, "cause");
+        if (eventLoop.inEventLoop()) {
+            close0(new CloseEvent(cause));
+        } else {
+            eventLoop.execute(() -> close0(new CloseEvent(cause)));
+        }
+    }
+
+    private void close0(CloseEvent closeEvent) {
+        assert eventLoop.inEventLoop();
+
+        if (state != State.OPEN) {
+            return;
+        }
+        state = State.CLOSED;
+
+        queue.add(closeEvent);
+        tryNotifyEvents();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return !isOpen() && !wroteAny;
+    }
+
+    @Override
+    public long demand() {
+        return demand;
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return completionFuture;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+        if (eventLoop.inEventLoop()) {
+            subscribe0(subscriber, executor, options);
+        } else {
+            eventLoop.execute(() -> subscribe0(subscriber, executor, options));
+        }
+    }
+
+    private void subscribe0(Subscriber<?> subscriber, EventExecutor executor,
+                            SubscriptionOption... options) {
+        assert eventLoop.inEventLoop();
+        if (subscription != null) {
+            notifySubscribeAndAbort(NoopSubscription.get(), subscriber,
+                                    abortedOrLate(subscription.subscriber()), executor);
+            return;
+        }
+        subscription = new SubscriptionImpl(this, subscriber, executor);
+        notifySubscribe0(subscription, subscriber, executor);
+        tryNotifyEvents();
+    }
+
+    @Override
+    public void abort() {
+        if (eventLoop.inEventLoop()) {
+            abort0(ABORTED_CLOSE);
+        } else {
+            eventLoop.execute(() -> abort0(ABORTED_CLOSE));
+        }
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        if (eventLoop.inEventLoop()) {
+            abort0(new CloseEvent(cause));
+        } else {
+            eventLoop.execute(() -> abort0(new CloseEvent(cause)));
+        }
+    }
+
+    private void abort0(CloseEvent closeEvent) {
+        assert eventLoop.inEventLoop();
```

Comment:
Remove redundant assert?

Reply:
I'm still trying to convince myself of the correctness of the implementation before starting to:

1) clean up test code, naming, redundant calls
2) dedup common logic with `[Default|Abstract]StreamMessage`

The `assert` statements across this file helps me make sure I can change state in a certain method when refactoring.
Let me clean up after this class is more organized 🙏 

(leaving this comment to let you know I'm not ignoring your review, I'll get to it eventually)

Reply:
(\+ thanks for taking the time to have a look - I know you're busy these days 🙇‍♂️ )

Reply:
> Let me clean up after this class is more organized 🙏

I see. No need to be in a hurry. Take your time. 🙂

Reply:
So I think this comment deserves some explanation.

I'm using `eventLoop.inEventLoop()` as a reminder whether it is OK to change state in this method or not.
The call itself requires a `Thread.currentThread()` followed by an equality comparison, which I don't think is very expensive.

Having said this, if you prefer I can remove the duplicated assertions.